modules/services/payment/src/main/kotlin/au/com/shiftyjelly/pocketcasts/payment/PaymentClient.kt

```diff
@@ -100,22 +98,23 @@ class PaymentClient @Inject constructor(
         dispatchConfirmation: Boolean,
     ) {
         if (purchase.state is PurchaseState.Purchased && pendingPurchases.add(purchase.state.orderId)) {
-            logger.info("Confirm purchase: $purchase")
-
-            val confirmResult = purchaseApprover.approve(purchase)
-                .flatMap { approvedPurchase ->
-                    if (!approvedPurchase.isAcknowledged) {
-                        dataSource.acknowledgePurchase(approvedPurchase)
-                    } else {
-                        PaymentResult.Success(approvedPurchase)
+            try {
+                forEachListener { onConfirmPurchase(purchase) }
+                val confirmResult = purchaseApprover.approve(purchase)
+                    .flatMap { approvedPurchase ->
+                        if (!approvedPurchase.isAcknowledged) {
+                            dataSource.acknowledgePurchase(approvedPurchase)
+                        } else {
+                            PaymentResult.Success(approvedPurchase)
+                        }
                     }
+                forEachListener { onPurchaseConfirmed(confirmResult) }
+                if (dispatchConfirmation) {
+                    _purchaseEvents.emit(confirmResult.toPurchaseResult())
                 }
-                .onSuccess { logger.info("Purchase confirmed: $it") }
-                .onFailure { code, message -> logger.warning("Failed to confirm purchase: $purchase. $code, $message") }
-            if (dispatchConfirmation) {
-                _purchaseEvents.emit(confirmResult.toPurchaseResult())
+            } finally {
+                pendingPurchases.remove(purchase.state.orderId)
```

Comment:
This accounts for any interruptions that might occur. It doesn't have a very meaningful difference for the production code as order IDs are unique there but it helps with testing and debug builds.